### PR TITLE
[git] Add the ability to fetch from different remote

### PR DESCRIPTION
**Pull request by droslean**
Add the ability to fetch from a different remote using the existing user and pass.

/cc @stevekuznetsov @alvaroaleman 
Signed-off-by: Nikolaos Moraitis <nmoraiti@redhat.com>

### CODE REVIEW

File: prow/git/git.go [hunk 1]
```diff
@@ -141,6 +141,43 @@ func (c *Client) unlockRepo(repo string) {
 	c.repoLocks[repo].Unlock()
 }
 
+// Init create an empty Git repository and returns a Repo
+func (c *Client) Init(organization, repository string) (*Repo, error) {
```
[the remark this answers is not in the document]
Comment (droslean) [reply prow/git/git.go]: Removed. Thanks

File: prow/git/git.go [hunk 2]
```diff
@@ -488,3 +525,38 @@ func (i *Repo) ShowRef(commitlike string) (string, error) {
 	}
 	return strings.TrimSpace(string(out)), nil
 }
+
+// Fetch runs a git fetch command for the specified remote. Pass the remote as org/repo format
+// since the actual URI will be constructed using the existing credentials.
+func (r *Repo) Fetch(remote, branch string) error {
```
[the remark this answers is not in the document]
Comment (alvaroaleman) [reply prow/git/git.go:494]: Yeah it seems there is no way to use different remotes in either the old or the new clients. I would probably rename it to `FetchFromDifferentRemote` or sth but generallu think this is fine.

Should have tests though and get implemented into the v2 client as well :)

File: prow/git/git.go [hunk 3]
```diff
@@ -488,3 +488,17 @@ func (i *Repo) ShowRef(commitlike string) (string, error) {
 	}
 	return strings.TrimSpace(string(out)), nil
 }
+
+// FetchFromRemote runs a git fetch command for the specified remote. Pass the remote as org/repo format
+// since the actual URI will be constructed using the existing credentials.
+func (r *Repo) FetchFromRemote(remote, branch string) error {
+	r.logger.Infof("Fetching from '%s (branch: %s)'.", remote, branch)
+	remoteURI := fmt.Sprintf("https://%s:%s@%s/%s", r.user, r.pass, r.host, remote)
```
Comment (alvaroaleman) [inline prow/git/git.go:496]: `org/repo` is not a `remote`, can you rename the var to `orgRepo` and mention in the godocs that this will always use https?

Comment (droslean) [reply prow/git/git.go]: addressed

File: prow/git/v2/interactor.go [hunk 4]
```diff
@@ -278,6 +282,32 @@ func (i *interactor) FetchRef(refspec string) error {
 	return nil
 }
 
+// FetchFromRemote fetches all update from a specific remote and branch and leaves it as FETCH_HEAD.
+func (i *interactor) FetchFromRemote(remote RemoteResolver, branch string) error {
+	r, err := remote()
```
Comment (stevekuznetsov) [inline prow/git/v2/interactor.go:287]: We only read this value once and don't store it -- why not just pass the URI as a string? What benefit is it to pass a resolver?

File: prow/git/v2/interactor.go [hunk 5]
```diff
@@ -278,6 +282,32 @@ func (i *interactor) FetchRef(refspec string) error {
...
+	if err != nil {
+		return fmt.Errorf("couldn't get remote: %v", err)
+	}
+
+	// Parse the remote's URI to censor any existing tokens.
+	u, err := url.Parse(r)
```
Comment (stevekuznetsov) [inline prow/git/v2/interactor.go:293]: IMO this should be on the user to do correctly by adding their tokens to the secret agent at startup

File: prow/git/v2/interactor.go [hunk 6]
```diff
@@ -278,6 +282,32 @@ func (i *interactor) FetchRef(refspec string) error {
...
+	if err != nil {
+		return fmt.Errorf("couldn't parse remote url: %v", err)
+	}
+
+	if token, ok := u.User.Password(); ok {
+		i.logger.Logger.SetFormatter(logrusutil.NewCensoringFormatter(logrus.StandardLogger().Formatter, func() sets.String {
```
Comment (stevekuznetsov) [inline prow/git/v2/interactor.go:299]: As written this code _overwrites_ the other formatters.

File: prow/git/git.go [hunk 7]
```diff
@@ -488,3 +488,18 @@ func (i *Repo) ShowRef(commitlike string) (string, error) {
 	}
 	return strings.TrimSpace(string(out)), nil
 }
+
+// FetchFromRemote runs a git fetch command for the specified remote.
+// Pass the destination as org/repo format since the actual URI will be constructed
+// using the existing credentials and will always use the https protocol.
+func (r *Repo) FetchFromRemote(orgRepo, branch string) error {
+	r.logger.Infof("Fetching from '%s (branch: %s)'.", orgRepo, branch)
+	remoteURI := fmt.Sprintf("https://%s:%s@%s/%s", r.user, r.pass, r.host, orgRepo)
```
Comment (alvaroaleman) [inline prow/git/git.go:497]: Interesting, didn't know this works without the `.git` suffix
/lgtm
/approve

Comment (alvaroaleman) [reply prow/git/git.go:497]: /hold

File: prow/git/v2/interactor.go (hunk 4 above)
Comment (droslean) [reply prow/git/v2/interactor.go:287]: Yea, we could do that. Since I already here parse the url.

File: prow/git/v2/interactor.go (hunk 6 above)
Comment (alvaroaleman) [reply prow/git/v2/interactor.go:299]: /lgtm cancel
Missed this

Comment (droslean) [reply prow/git/v2/interactor.go:299]: We can remove this if we rely on user to censor the secrets.

File: prow/git/git.go (hunk 7 above)
Comment (alvaroaleman) [reply prow/git/git.go:497]: This will break if user or password are unset, no? Do we generally assume that they are always set?

Comment (droslean) [reply prow/git/git.go:497]: Adding a check for this won't hurt. Addressing...

**Comment by droslean** [comment]
Oops, messed up the rebase.

File: prow/git/git.go (hunk 7 above)
Comment (droslean) [reply prow/git/git.go]: Addressed

**Pull request merged**